huaweicloud/services/waf/resource_huaweicloud_waf_certificate.go

```diff
@@ -67,85 +68,104 @@ func ResourceWafCertificateV1() *schema.Resource {
 	if err != nil {
-		return fmtp.Errorf("error creating HuaweiCloud WAF Client: %s", err)
+		return diag.Errorf("error creating WAF client: %s", err)
 	}
 
-	epsID := config.GetEnterpriseProjectID(d)
-	n, err := certificates.GetWithEpsID(wafClient, d.Id(), epsID).Extract()
+	epsID := cfg.GetEnterpriseProjectID(d)
+	n, err := certificates.GetWithEpsID(client, d.Id(), epsID).Extract()
 	if err != nil {
-		return common.CheckDeleted(d, err, "Error obtain WAF certificate information")
+		return common.CheckDeletedDiag(d, err, "error retrieving WAF certificate")
 	}
 
 	expires := time.Unix(int64(n.ExpireTime/1000), 0).UTC().Format("2006-01-02 15:04:05 MST")
-
-	d.Set("region", config.GetRegion(d))
-	d.Set("name", n.Name)
-	d.Set("expiration", expires)
-
-	return nil
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("name", n.Name),
+		d.Set("expiration", expires),
+	)
+	return diag.FromErr(mErr.ErrorOrNil())
 }
 
-func resourceWafCertificateV1Update(d *schema.ResourceData, meta interface{}) error {
-	config := meta.(*config.Config)
-	wafClient, err := config.WafV1Client(config.GetRegion(d))
+func resourceWafCertificateV1Update(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	client, err := cfg.WafV1Client(cfg.GetRegion(d))
 	if err != nil {
-		return fmtp.Errorf("error creating HuaweiCloud WAF Client: %s", err)
+		return diag.Errorf("error creating WAF client: %s", err)
 	}
-
-	updateOpts := certificates.UpdateOpts{
-		Name:                d.Get("name").(string),
-		EnterpriseProjectId: config.GetEnterpriseProjectID(d),
+	if d.HasChange("name") {
+		updateOpts := certificates.UpdateOpts{
+			Name:                d.Get("name").(string),
+			EnterpriseProjectId: cfg.GetEnterpriseProjectID(d),
+		}
+
+		_, err = certificates.Update(client, d.Id(), updateOpts).Extract()
+		if err != nil {
+			return diag.Errorf("error updating WAF certificate name: %s", err)
+		}
 	}
 
-	_, err = certificates.Update(wafClient, d.Id(), updateOpts).Extract()
-	if err != nil {
-		return fmtp.Errorf("error updating WAF Certificate: %w", err)
+	if d.HasChange("certificate") {
+		oldRaw, newRaw := d.GetChange("certificate")
+		oldVal := strings.TrimSpace(oldRaw.(string))
+		newVal := strings.TrimSpace(newRaw.(string))
+		if oldVal != newVal {
+			// `name` is required for updating operation.
+			// `certificate` and `private_key` must be specified together.
+			// Ignore the updating operation when field `certificate` does not change.
+			updateOpts := certificates.UpdateOpts{
+				Name:                d.Get("name").(string),
+				EnterpriseProjectId: cfg.GetEnterpriseProjectID(d),
+				Content:             newVal,
+				Key:                 strings.TrimSpace(d.Get("private_key").(string)),
+			}
+			_, err = certificates.Update(client, d.Id(), updateOpts).Extract()
+			if err != nil {
+				return diag.Errorf("error updating WAF certificate content or key: %s", err)
+			}
+		}
 	}
```

Comment:
```suggestion
        updateOpts := certificates.UpdateOpts{}
        if d.HasChange("name") {
		updateOpts .Name = d.Get("name").(string)
		updateOpts .EnterpriseProjectId= cfg.GetEnterpriseProjectID(d)
	}
	if d.HasChange("certificate") {
		oldRaw, newRaw := d.GetChange("certificate")
		oldVal := strings.TrimSpace(oldRaw.(string))
		newVal := strings.TrimSpace(newRaw.(string))
		if oldVal != newVal {
			// `name` is required for updating operation.
			// `certificate` and `private_key` must be specified together.
			// Ignore the updating operation when field `certificate` does not change.
			updateOpts .Name = d.Get("name").(string)
			updateOpts .EnterpriseProjectId= cfg.GetEnterpriseProjectID(d)
			updateOpts .Content= newVal
			updateOpts .Key= strings.TrimSpace(d.Get("private_key").(string))
		}
	}
	// if the `name` is not empty, it indicates that at least one of name and certificate should be updated
	if len(updateOpts .Name) > 0{
	  _, err = certificates.Update(client, d.Id(), updateOpts).Extract()
			if err != nil {
				return diag.Errorf("error updating WAF certificate content or key: %s", err)
			}
	}
```

Reply:
Has changed another way. Please check the new code.